snuba/migrations/snuba_migrations/metrics/0002_metrics_sets.py

```diff
@@ -0,0 +1,160 @@
+from typing import Sequence
+
+from snuba.clickhouse.columns import (
+    AggregateFunction,
+    Array,
+    Column,
+    DateTime,
+    Nested,
+    UInt,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.storages.tags_hash_map import INT_TAGS_HASH_MAP_COLUMN
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+
+sets_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("metric_id", UInt(64)),
+    Column("granularity", UInt(32)),
+    Column("timestamp", DateTime()),
+    Column("retention_days", UInt(16)),
+    Column("value", AggregateFunction("uniqCombined64", [UInt(64)])),
+]
+
+sets_mv_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("metric_id", UInt(64)),
+    Column("granularity", UInt(32)),
+    Column("timestamp", DateTime()),
+    Column("retention_days", UInt(16)),
+    Column("tags", Nested([Column("key", UInt(64)), Column("value", UInt(64))]),),
+    Column("value", AggregateFunction("uniqCombined64", [UInt(64)])),
+]
+
+
+matview_statement = """
+SELECT
+    org_id,
+    project_id,
+    metric_id,
+    60 as granularity,
+    toStartOfMinute(timestamp) as timestamp,
+    retention_days,
+    tags.key,
+    tags.value,
+    uniqCombined64State(arrayJoin(set_values)) as value
```

Comment:
What does the `State` modifier do here?

Reply:
The materialization statement generates an aggregation state which is stored in the aggregated table.
An aggregation state is a binary blob that contain a data sketch used to calculate the result of tthe aggregation function.
Aggregation states iniclude enough data to be merged with others but they do not contain all the original data so they remain small.
When we query we run `uniqCombined64Merge` over one or multiple aggregation states to generates the result (which is the number of unique users for example).

An example of aggregation state is a HyperLogLog digest. `uniqHLLState` would generate the digest from the raw data. `uniqHLLMerge` would extract the number of unique elements from the digest.

https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/combinators/#agg-functions-combinator-state